src/raftstore/store/fsm/router.rs

```diff
@@ -0,0 +1,627 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This is the core implementation of a batch system. Generally there will be two
+//! different kind of FSMs in TiKV's FSM system. One is normal FSM, which usually
+//! represents a peer, the other is control FSM, which usually represents something
+//! that controls how the former is created or metrics are collected.
+
+// TODO: remove this
+#![allow(dead_code)]
+
+use crossbeam::channel::{SendError, TrySendError};
+use std::borrow::Cow;
+use std::cell::Cell;
+use std::ptr;
+use std::sync::atomic::{AtomicPtr, AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex};
+use util::collections::HashMap;
+use util::mpsc;
+use util::Either;
+
+// The FSM is notified.
+const NOTIFYSTATE_NOTIFIED: usize = 0;
```

Comment:
How about
```
#[repr(C)]
enum NotifyState
```
?

Reply:
Using enum will require additional conversion, which makes code hard to read.

Reply:
I see. We need to store it in a atomic.